File: app/src/ultimate/java/com/clockmods/widget/update/WidgetUpdateCoordinator.java
package com.clockmods.widget.update;
import android.appwidget.*;
import android.content.*;
import com.clockmods.widget.model.*;
import com.clockmods.widget.store.WidgetConfigStore;
import com.clockmods.widget.provider.*;
import com.clockmods.widget.render.WidgetRemoteViewsFactory;
import java.util.*;
import java.util.concurrent.*;
public final class WidgetUpdateCoordinator {
    private static final ExecutorService EXECUTOR=Executors.newSingleThreadExecutor();
    private static final Class<?>[] PROVIDERS={DigitalClockWidgetProvider.class,AnalogClockWidgetProvider.class,WeatherClockWidgetProvider.class,CalendarWidgetProvider.class};
    private WidgetUpdateCoordinator() { }
    public static void execute(Runnable work) { EXECUTOR.execute(work); }
    public static WidgetKind kindFor(Context c,int id) {
        AppWidgetProviderInfo info=AppWidgetManager.getInstance(c).getAppWidgetInfo(id);
        if(info==null || !c.getPackageName().equals(info.provider.getPackageName())) return null;
        for(int i=0;i<PROVIDERS.length;i++) if(PROVIDERS[i].getName().equals(info.provider.getClassName())) return WidgetKind.values()[i];
        return null;
    }
    public static List<WidgetConfig> active(Context c) {
        List<WidgetConfig> all=new ArrayList<>(); WidgetConfigStore store=new WidgetConfigStore(c);
        for(int i=0;i<PROVIDERS.length;i++) for(int id:AppWidgetManager.getInstance(c).getAppWidgetIds(new ComponentName(c,PROVIDERS[i]))) {
            WidgetConfig config=store.getOrDefault(id,WidgetKind.values()[i]);
            if(!store.contains(id)) store.save(config);
            all.add(config);
        }
        return all;
    }
    public static int weatherCount(Context c) { return AppWidgetManager.getInstance(c).getAppWidgetIds(new ComponentName(c,WeatherClockWidgetProvider.class)).length; }
    public static void updateOne(Context c,int id) {
        WidgetKind kind=kindFor(c,id); if(kind==null) return;
        AppWidgetManager.getInstance(c).updateAppWidget(id,WidgetRemoteViewsFactory.createResponsive(c,id,kind));
    }
    public static void updateAll(Context c) {
        for(WidgetConfig config:active(c)) updateOne(c,config.appWidgetId);
        reconcile(c);
    }
    public static void updateWeatherWidgets(Context c) {
        for(int id:AppWidgetManager.getInstance(c).getAppWidgetIds(new ComponentName(c,WeatherClockWidgetProvider.class))) updateOne(c,id);
    }
    public static void reconcile(Context c) {
        List<WidgetConfig> configs=active(c); Set<Integer> ids=new HashSet<>(); int weather=0;
        for(WidgetConfig config:configs) { ids.add(config.appWidgetId); if(config.kind==WidgetKind.WEATHER) weather++; }
        WidgetConfigStore store=new WidgetConfigStore(c);
        for(WidgetConfig config:store.getAll()) if(!ids.contains(config.appWidgetId)) store.delete(config.appWidgetId);
        WidgetRefreshWorker.schedule(c,weather); WidgetMidnightScheduler.schedule(c,configs);
    }
}

File: app/src/ultimate/java/com/clockmods/widget/update/WidgetUpdateCoordinator.kt
package com.clockmods.widget.update

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import com.clockmods.widget.model.WidgetConfig
import com.clockmods.widget.model.WidgetKind
import com.clockmods.widget.provider.AnalogClockWidgetProvider
import com.clockmods.widget.provider.CalendarWidgetProvider
import com.clockmods.widget.provider.DigitalClockWidgetProvider
import com.clockmods.widget.provider.WeatherClockWidgetProvider
import com.clockmods.widget.render.WidgetRemoteViewsFactory
import com.clockmods.widget.store.WidgetConfigStore
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

object WidgetUpdateCoordinator {
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private val providers = arrayOf(
        DigitalClockWidgetProvider::class.java,
        AnalogClockWidgetProvider::class.java,
        WeatherClockWidgetProvider::class.java,
        CalendarWidgetProvider::class.java,
    )

    @JvmStatic fun execute(work: Runnable) {
        executor.execute(work)
    }

    @JvmStatic fun kindFor(context: Context, id: Int): WidgetKind? {
        val info = AppWidgetManager.getInstance(context).getAppWidgetInfo(id)
        if (info == null || context.packageName != info.provider.packageName) return null
        for (index in providers.indices) {
            if (providers[index].name == info.provider.className) return WidgetKind.values()[index]
        }
        return null
    }

    @JvmStatic fun active(context: Context): List<WidgetConfig> {
        val all = ArrayList<WidgetConfig>()
        val store = WidgetConfigStore(context)
        val manager = AppWidgetManager.getInstance(context)
        for (index in providers.indices) {
            val kind = WidgetKind.values()[index]
            for (id in manager.getAppWidgetIds(ComponentName(context, providers[index]))) {
                val config = store.getOrDefault(id, kind)
                if (!store.contains(id)) store.save(config)
                all += config
            }
        }
        return all
    }

    @JvmStatic fun weatherCount(context: Context): Int = AppWidgetManager.getInstance(context)
        .getAppWidgetIds(ComponentName(context, WeatherClockWidgetProvider::class.java)).size

    @JvmStatic fun updateOne(context: Context, id: Int) {
        val kind = kindFor(context, id) ?: return
        AppWidgetManager.getInstance(context).updateAppWidget(id, WidgetRemoteViewsFactory.createResponsive(context, id, kind))
    }

    @JvmStatic fun updateAll(context: Context) {
        active(context).forEach { updateOne(context, it.appWidgetId) }
        reconcile(context)
    }

    @JvmStatic fun updateWeatherWidgets(context: Context) {
        val manager = AppWidgetManager.getInstance(context)
        manager.getAppWidgetIds(ComponentName(context, WeatherClockWidgetProvider::class.java))
            .forEach { updateOne(context, it) }
    }

    @JvmStatic fun reconcile(context: Context) {
        val configs = active(context)
        val ids = configs.mapTo(HashSet()) { it.appWidgetId }
        val weather = configs.count { it.kind == WidgetKind.WEATHER }
        val store = WidgetConfigStore(context)
        store.getAll().filterNot { ids.contains(it.appWidgetId) }.forEach { store.delete(it.appWidgetId) }
        WidgetRefreshWorker.schedule(context, weather)
        WidgetMidnightScheduler.schedule(context, configs)
    }
}
